// backend/src/models/user.model.mjs
import mongoose from "mongoose";

const userRoles = [
    "Super Admin",
    "Hospital Admin",
    "Doctor",
    "Nurse",
    "Receptionist",
    "Lab Technician",
    "Pharmacist",
    "Accountant",
    "Patient",
];

const userSchema = new mongoose.Schema(
    {
        hospitalId: {type: mongoose.Schema.Types.ObjectId, ref: "Hospital"},
        email: {type: String, required: true, unique: true},
        passwordHash: {type: String, required: true},
        role: {type: String, enum: userRoles, required: true},
        firstName: {type: String, required: true},
        lastName: {type: String, required: true},
        phone: {type: String},
        avatarUrl: {type: String},
        isEmailVerified: {type: Boolean, default: false},
        isActive: {type: Boolean, default: true},
        lastLoginAt: {type: Date},
    },
    {timestamps: true},
);

userSchema.index({email: 1}, {unique: true});
userSchema.index({hospitalId: 1, role: 1});
userSchema.index({hospitalId: 1, isActive: 1});

const User = mongoose.model("User", userSchema);

export default User;
